File: ParaView/GUI/Widgets/Examples/Cxx/WidgetsTour/Widgets/vtkKWCornerAnnotationEditor.cxx
#include "vtkCornerAnnotation.h"
#include "vtkImageData.h"
#include "vtkImageViewer2.h"
#include "vtkKWApplication.h"
#include "vtkKWCornerAnnotationEditor.h"
#include "vtkKWFrame.h"
#include "vtkKWRenderWidget.h"
#include "vtkKWWindow.h"
#include "vtkRenderWindow.h"
#include "vtkRenderWindowInteractor.h"
#include "vtkXMLImageDataReader.h"

#include "vtkKWWidgetsTourExample.h"

class vtkKWCornerAnnotationEditorItem : public KWWidgetsTourItem
{
public:
  virtual int GetType();
  virtual void Create(vtkKWWidget *parent, vtkKWWindow *win);

  vtkKWCornerAnnotationEditorItem();
  virtual ~vtkKWCornerAnnotationEditorItem();

protected:
  vtkKWRenderWidget *cae_renderwidget;
  vtkXMLImageDataReader *cae_reader;
  vtkImageViewer2 *cae_viewer;
  vtkKWCornerAnnotationEditor *cae_anno_editor;
};

void vtkKWCornerAnnotationEditorItem::Create(vtkKWWidget *parent, vtkKWWindow *win)
{
  vtkKWApplication *app = parent->GetApplication();

  // -----------------------------------------------------------------------

  // Create a render widget
  // Set the corner annotation visibility

  this->cae_renderwidget = vtkKWRenderWidget::New();
  this->cae_renderwidget->SetParent(parent);
  this->cae_renderwidget->Create();
  this->cae_renderwidget->CornerAnnotationVisibilityOn();

  app->Script("pack %s -side right -fill both -expand y -padx 0 -pady 0", 
              this->cae_renderwidget->GetWidgetName());

  // -----------------------------------------------------------------------

  // Create a volume reader

  this->cae_reader = vtkXMLImageDataReader::New();
  this->cae_reader->SetFileName(
    vtkKWWidgetsTourExample::GetPathToExampleData(app, "head100x100x47.vti"));

  // Create an image viewer
  // Use the render window and renderer of the renderwidget

  this->cae_viewer = vtkImageViewer2::New();
  this->cae_viewer->SetRenderWindow(this->cae_renderwidget->GetRenderWindow());
  this->cae_viewer->SetRenderer(this->cae_renderwidget->GetRenderer());
  this->cae_viewer->SetInput(this->cae_reader->GetOutput());
  this->cae_viewer->SetupInteractor(
    this->cae_renderwidget->GetRenderWindow()->GetInteractor());

  // Reset the window/level and the camera

  this->cae_reader->Update();
  double *range = this->cae_reader->GetOutput()->GetScalarRange();
  this->cae_viewer->SetColorWindow(range[1] - range[0]);
  this->cae_viewer->SetColorLevel(0.5 * (range[1] + range[0]));

  this->cae_renderwidget->ResetCamera();

  // -----------------------------------------------------------------------

  // The corner annotation has the ability to parse "tags" and fill
  // them with information gathered from other objects.
  // For example, let's display the slice and window/level in one corner
  // by connecting the corner annotation to our image actor and
  // image mapper

  vtkCornerAnnotation *ca = this->cae_renderwidget->GetCornerAnnotation();
  ca->SetImageActor(this->cae_viewer->GetImageActor());
  ca->SetWindowLevel(this->cae_viewer->GetWindowLevel());
  ca->SetText(2, "<slice>");
  ca->SetText(3, "<window>\n<level>");
  ca->SetText(1, "Hello, World!");

  // -----------------------------------------------------------------------

  // Create a corner annotation editor
  // Connect it to the render widget
  
  this->cae_anno_editor = vtkKWCornerAnnotationEditor::New();
  this->cae_anno_editor->SetParent(parent);
  this->cae_anno_editor->Create();
  this->cae_anno_editor->SetRenderWidget(this->cae_renderwidget);

  app->Script("pack %s -side left -anchor nw -expand n -padx 2 -pady 2", 
              this->cae_anno_editor->GetWidgetName());
}

int vtkKWCornerAnnotationEditorItem::GetType()
{
  return KWWidgetsTourItem::TypeVTK;
}

KWWidgetsTourItem* vtkKWCornerAnnotationEditorEntryPoint()
{
  return new vtkKWCornerAnnotationEditorItem();
}

vtkKWCornerAnnotationEditorItem::vtkKWCornerAnnotationEditorItem()
{
  this->cae_anno_editor = NULL;
  this->cae_reader = NULL;
  this->cae_renderwidget = NULL;
  this->cae_viewer = NULL;
}

vtkKWCornerAnnotationEditorItem::~vtkKWCornerAnnotationEditorItem()
{
  if (this->cae_anno_editor)
    {
    this->cae_anno_editor->Delete();
    }
  if (this->cae_reader)
    {
    this->cae_reader->Delete();
    }
  if (this->cae_renderwidget)
    {
    this->cae_renderwidget->Delete();
    }
  if (this->cae_viewer)
    {
    this->cae_viewer->Delete();
    }
}

File: GUI/Widgets/Examples/Python/WidgetsTour/Widgets/VTK/vtkKWCornerAnnotationEditor.py
from vtk import vtkCornerAnnotation
from vtk import vtkImageData
from vtk import vtkImageViewer2
from kwwidgets import vtkKWApplication
from kwwidgets import vtkKWCornerAnnotationEditor
from kwwidgets import vtkKWFrame
from kwwidgets import vtkKWRenderWidget
from kwwidgets import vtkKWWindow
from vtk import vtkRenderWindow
from vtk import vtkRenderWindowInteractor
from vtk import vtkXMLImageDataReader

import os



def vtkKWCornerAnnotationEditorEntryPoint(parent, win):

    app = parent.GetApplication()
    
    # -----------------------------------------------------------------------
    
    # Create a render widget
    # Set the corner annotation visibility
    
    cae_renderwidget = vtkKWRenderWidget()
    cae_renderwidget.SetParent(parent)
    cae_renderwidget.Create()
    cae_renderwidget.CornerAnnotationVisibilityOn()
    
    app.Script("pack %s -side right -fill both -expand y -padx 0 -pady 0",
        cae_renderwidget.GetWidgetName())
    
    # -----------------------------------------------------------------------
    
    # Create a volume reader
    
    cae_reader = vtkXMLImageDataReader()
    cae_reader.SetFileName(os.path.join(
        os.path.dirname(os.path.abspath(__file__)),
        "..", "..", "..", "..", "Data", "head100x100x47.vti"))
    
    # Create an image viewer
    # Use the render window and renderer of the renderwidget
    
    cae_viewer = vtkImageViewer2()
    cae_viewer.SetRenderWindow(cae_renderwidget.GetRenderWindow())
    cae_viewer.SetRenderer(cae_renderwidget.GetRenderer())
    cae_viewer.SetInput(cae_reader.GetOutput())
    cae_viewer.SetupInteractor(
        cae_renderwidget.GetRenderWindow().GetInteractor())
    
    # Reset the window/level and the camera
    
    cae_reader.Update()
    range = cae_reader.GetOutput().GetScalarRange()
    cae_viewer.SetColorWindow(range[1] - range[0])
    cae_viewer.SetColorLevel(0.5 * (range[1] + range[0]))
    
    cae_renderwidget.ResetCamera()
    
    # -----------------------------------------------------------------------
    
    # The corner annotation has the ability to parse "tags" and fill
    # them with information gathered from other objects.
    # For example, let's display the slice and window/level in one corner
    # by connecting the corner annotation to our image actor and
    # image mapper
    
    ca = cae_renderwidget.GetCornerAnnotation()
    ca.SetImageActor(cae_viewer.GetImageActor())
    ca.SetWindowLevel(cae_viewer.GetWindowLevel())
    ca.SetText(2, "<slice>")
    ca.SetText(3, "<window>\n<level>")
    ca.SetText(1, "Hello, World!")
    
    # -----------------------------------------------------------------------
    
    # Create a corner annotation editor
    # Connect it to the render widget
    
    cae_anno_editor = vtkKWCornerAnnotationEditor()
    cae_anno_editor.SetParent(parent)
    cae_anno_editor.Create()
    cae_anno_editor.SetRenderWidget(cae_renderwidget)
    
    app.Script("pack %s -side left -anchor nw -expand n -padx 2 -pady 2",
        cae_anno_editor.GetWidgetName())
    
    return "TypeVTK"
